Feature/KAS-3024 Pub v3 Reports: Update columns

mdv-professional opened the pull request:
[Jira](https://kanselarij.atlassian.net/browse/KAS-3024)

**Vragen**
Voor de `received-date` wordt onder translations en onder proofs een andere logica gebruikt.
[translations](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/development/app/pods/publications/publication/translations/requests/controller.js#L46)
[proofs](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/development/app/pods/publications/publication/proofs/requests/controller.js#L98)
Is dit misschien een vergissing?

Commit:
KAS-3024 add and reorder columns
> update columns [pub v3]

Commit:
KAS-3024 page count
> update columns [pub v3: reports]

Commit:
KAS-3024 page count | fix: update reference loadDecision > loadData
> update columns [pub v3: reports]

Commit:
KAS-3024 fix: test absence of start date in translation-subcase when saving a translation request
> update columns [pub v3: reports]

Commit:
KAS-3024 page count refactor to method
> update columns [pub v3: reports]

Commit:
KAS-3024 page count (comment) toArray() workaround
> update columns [pub v3: reports]

Commit:
KAS-3024 translation request date
> update columns [pub v3: reports]

Commit:
KAS-3024 translation request date (cleanup) remove TODO
> update columns [pub v3: reports]

Commit:
KAS-3024 proof request date
> update columns [pub v3: reports]

Commit:
KAS-3024 proof request date (cleanup) remove log
> update columns [pub v3: reports]

Commit:
KAS-3024 proof received date
> update columns [pub v3: reports]

Commit:
KAS-3024 proof request date (simplify)
> update columns [pub v3: reports]

Commit:
KAS-3024 (refactor) split loadData in loadData & loadDecision method
> update columns [pub v3: reports]

Commit:
KAS-3024 (fix) findRecord > queryOne
> update columns [pub v3: reports]

Commit:
KAS-3024 proof received date (cleanup) remove TODO
> update columns [pub v3: reports]

Commit:
KAS-3024 showByDefault === false for new columns
> update columns [pub v3: reports]

Commit:
KAS-3024 test:  publicationReceivedDate > proofReceivedDate
> update columns [pub v3: reports]

Commit:
KAS-3024 test:  publicationReceivedDate > proofReceivedDate
> update columns [pub v3: reports]

Commit:
KAS-3024 test:  publicationReceivedDate > proofReceivedDate
> update columns [pub v3: reports]

Commit:
Merge branch 'feature/KAS-3024-pub-feedback--update-columns' of https://github.com/kanselarij-vlaanderen/kaleidos-frontend into feature/KAS-3024-pub-reports--update-columns

Commit:
KAS-3024 remove sort on page count
> update columns [pub v3: reports]

Commit:
KAS-3024 translations: remove unused key last-edited-by
> update columns [pub v3: reports]

Commit:
KAS-3024 sort on numac number
> update columns [pub v3: reports]

mdv-professional commented on cypress/integration/unit/publication-overview.spec.js:
```diff
@@ -92,26 +92,26 @@ context('Publications overview tests', () => {
       'numacNumber',
+      'shortTitle',
+      'comment',
+      // 'pages', // hidden by default
```
De testcode voor kolommen hidden by default, moet vermoedelijk nog worden aangepast. @MathiasZabot

mdv-professional replied:
OK

mdv-professional commented on app/config/publications/overview-table-columns.js:
```diff
@@ -1,4 +1,25 @@
+  {
+    keyName: 'speedProcedure',
+    translationKey: 'publications-table-speed-procedure',
+    translationKeySmall: 'publications-table-speed-procedure-small',
```
Deze keys hebben nog hun oude namen. Het is wat verwarrend, maar dan blijft de configuratie in localStorage correct.

erikap replied:
Aangezien publicaties nog niet in productie gebruikt worden, zou ik me niet aantrekken om hier een breaking change te doen indien nodig.

mdv-professional commented on app/components/publications/publications-table-row.js:
```diff
@@ -25,6 +29,52 @@ export default class PublicationsPublicationsTableRowComponent extends Component
+  @task
+  *loadData() {
+    // using queryOne instead of findRecord: findRecord causes included records to be fetched again
+    const publicationFlow = yield this.store.queryOne('publication-flow', {
```
Mijn voorstel is om loadDecisions in dezelfde query te verwerken om extra requests te vermijden. Wanneer de conversie heeft gelopen gaat de publicaties-pagina heel traag door erg veel requests.

erikap replied:
Mag je proberen van mij. Bij toevoegen van `include` in het achterhoofd houden dat dat ervoor kan zorgen dat een cache-entry voor de request sneller geinvalideerd zal worden. De cache clearing is zeer naief/agressief: elke wijziging aan een `decision` zal alle requests met een `include=decision` clearen. Er wordt door de cache dus niet gecheckt of de gewijzigde `decision` diegene is die included is.

Nu zie ik dat in de huidige request van `publication-flow` al andere resources geinclude zijn waarvan de kans veel groter is dat die de cache-entry zullen invalideren dan de extra include van `decision`. Voeg dus maar toe.

mdv-professional commented on app/config/publications/overview-table-columns.js:
```diff
@@ -1,4 +1,25 @@
+    keyName: 'numacNumber',
+    translationKey: 'publications-table-numacnummer-bs',
+    translationKeySmall: 'publications-table-numacnummer-bs-small',
+    sortKey: 'numac-numbers.id-name,-created',
```
Sortering zou hier ook rekening kunnen houden met het versienummer.

erikap replied:
Ik vermoed dat je deze comment bij `publicationNumber` bedoelde ipv bij `numacNumber` aangezien er voor numac nummer geen versie ingevuld wordt?

Het veld met de versie is een string. Daarop sorteren geeft niet het gewenste effect voor cijfers of voor termen als 'bis', 'tris', ... De 2e sortering op `-created` is daarom toegevoegd in de veronderstelling dat nieuwere versies later gemaakt worden. Dat is zeker niet foutloos, maar een best effort.

mdv-professional commented on app/config/publications/overview-table-columns.js:
```diff
@@ -12,6 +33,13 @@ export default [
+    keyName: 'pages',
+    translationKey: 'publications-table-pages',
+    translationKeySmall: 'publications-table-pages-small',
+    // no sort: is aggregated in frontend
```
Sortering hiervan is onmogelijk. Dit is niet volgens het design.

erikap replied:
Inderdaad, niet mogelijk. Indien dat een must-have zou blijken voor de gebruikers (wat ik betwijfel) moet er een nieuwe story voorzien worden.

mdv-professional commented on app/pods/publications/publication/translations/documents/controller.js:
```diff
@@ -128,7 +128,7 @@ export default class PublicationsPublicationTranslationsDocumentController exten
   @task
   *saveTranslationRequest(translationRequest) {
     const now = new Date();
-    if (this.translationSubcase.startDate === null) {
```
Bugfix: startDate is undefined indien niet gestart.

Commit:
KAS-3024 correct date sort keys
> update columns [pub v3: reports]

Commit:
KAS-3024 remove obsolete columns translations + proofs
> update columns [pub v3: reports]

mdv-professional commented on app/components/publications/publications-table-row.hbs:
```diff
@@ -44,97 +86,98 @@ request the whole row and not only the arrow at the end to be clickable }}
       class="auk-u-text-nowrap"
     >
-      {{#if @publicationFlow.publicationSubcase.targetEndDate}}
+      {{#if @publicationFlow.publicationSubcase.receivedDate}}
```
Is het zinvol om hier de vroegste datum te nemen? Voor de opvolging van publicatieflows lijkt het me handiger om de datum leeg te laten wanneer er nog een aanvraag open staat.

erikap replied:
Te bekijken met de gebruikers maar dat zou ik uit deze PR houden, want gaat ook impact hebben op de logica bij het maken van nieuwe aanvragen.

Commit:
KAS-3024 translations: spelfout
> update columns [pub v3: reports]

Commit:
KAS-3024 translations: remove unused column names
> update columns [pub v3: reports]

ghost commented:
Ik ben hier nog iets vergeten: de waarschuwingen dat iets over tijd is.

Commit:
remove class "auk-u-text-align--center" of pages column in publications/overview
KAS-3024 pub: v3: reports: columns

Commit:
add "Datum verstreken" message for translation due date & publication due date publications/overview
KAS-3024 pub: v3: reports: columns

Commit:
update & add date comments of translation-subcase & publication-subcase
KAS-3024 pub: v3: reports: columns

Commit:
(lint-fix)
KAS-3024 pub: v3: reports: columns

erikap commented:
> **Vragen** Voor de `received-date` wordt onder translations en onder proofs een andere logica gebruikt. [translations](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/development/app/pods/publications/publication/translations/requests/controller.js#L46) [proofs](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/development/app/pods/publications/publication/proofs/requests/controller.js#L98) Is dit misschien een vergissing?


Ik zou in beide gevallen voor `now` gaan dan.

erikap commented on app/components/publications/publications-table-row.hbs:
```diff
@@ -7,6 +7,36 @@ request the whole row and not only the arrow at the end to be clickable }}
 >
   {{! template-lint-enable no-invalid-interactive }}
+  {{#if @tableColumnDisplayOptions.speedProcedure}}
+  <td
```
Indendatie

erikap commented on app/components/publications/publications-table-row.hbs:
```diff
@@ -7,6 +7,36 @@ request the whole row and not only the arrow at the end to be clickable }}
+    {{/if}}
+  </td>
+  {{/if}}
+    {{#if @tableColumnDisplayOptions.publicationNumber}}
```
Indendatie

erikap commented on app/components/publications/publications-table-row.hbs:
```diff
@@ -158,48 +210,52 @@ request the whole row and not only the arrow at the end to be clickable }}
+      class="auk-u-text-nowrap"
+    >
+      {{#if @publicationFlow.publicationSubcase.dueDate}}
+        {{!Ember seems not to await async getter}}
```
Dat klopt. Op sommige plaatsen in de code wordt hiervoor gebruik gemaakt van een [PromiseObject](https://api.emberjs.com/ember-data/3.26.0/classes/PromiseObject), waardoor de view wel update. Maar we trachten dit patroon te vermijden. Het verleden heeft geleerd dat dat leidt tot een oncontroleerbare stroom van network requests, zeker wanneer getters op elkaar gaan dependen.

erikap commented on app/components/publications/publications-table-row.hbs:
```diff
@@ -158,48 +210,52 @@ request the whole row and not only the arrow at the end to be clickable }}
+    >
+      {{#if @publicationFlow.publicationSubcase.dueDate}}
+        {{!Ember seems not to await async getter}}
+        <div class="{{if (await this.isPublicationToLate) "auk-u-text-bold"}}">
```
Ik denk dat je deze `await`s kan vermijden door de functies die je nu in `publication-utils.js` hebt gedefineerd op de Ember Data models toe te voegen.

Vb. op publication-subcase
```javascript
get isTooLate() {
    return moment(this.dueDate).isBefore(Date.now(), 'day')
}
```

Op die manier kan je in de template gewoon `@publicationFlow.publicationSubcase.isTooLate` gebruiken.

Gelijkaardige redenering voor de `isClosed` op publication-status.

mdv-professional replied:
Waarom ik het niet zo gedaan heb is omdat dit logica is die businesskennis vereist. Die kan volgens mij best maar op 1 plaats uitgeschreven staan. Als ik de logica in de models stak, stak de kennis deels in `publication-subcase`, deels in `publication-status` en deels in de template.
Om die awaits te vermijden, stel ik `@tracked`-properties voor.

erikap replied:
`@tracked`-properties zijn ook goed voor mij.

erikap replied:
Nu ik er nog eens naar kijk, lijkt de 'businesskennis' die nodig is mij hier eerder verwaarloosbaar en is het eerder logisch om de derived properties toe te voegen aan de ember-data models. 

- Voor publication-status is er intussen in de development branch vanuit een PR een derived property `isFinished` toegevoegd die je kan gebruiken.
- Voor de translation-subcase en publication-subcase modellen met een attribute `dueDate` lijkt een derived property `isOverdue` niet geheel onlogisch.

erikap reviewed (commented):
De async publication-utils zou ik eens herbekijken om op de ember-data models toe te voegen. Dat gaat de code volgens mij eenvoudiger maken. Voor de rest is deze PR ok voor mij.

Commit:
include decisions in a row's loadData request publications/overview
KAS-3024/PR pub: v3: reports: columns

Commit:
set end date of translationActivity to now
KAS-3024/PR pub: v3: reports: columns

Commit:
(lint-fix) publications-table-row
KAS-3024/PR pub: v3: reports: columns

Commit:
(rename) update column names in publications/overview
KAS-3024/PR pub: v3: reports: columns

Commit:
(refactor) move publication flow business knowledge from utils to service
KAS-3024/PR pub: v3: reports: columns

ghost commented:
Ik heb @erikap 's aanpassingen verwerkt behalve 1: het herbekijken van het async-getter-pattern. @MikiDi was met iets gelijkaardigs bezig i.v.m. `sign-flow`-models.

Commit:
tests: disable publication tests temporarily while new designs are being implemented
KAS-3024/PR pub: v3: reports: columns

Commit:
Merge branch 'development' into feature/KAS-3024-pub-feedback--update-columns

Commit:
replace includeZero parameter from if in publication-table-row
reason: not supported by new Ember version
KAS-3024/PR pub: v3: reports: columns

Commit:
(rename) pages to pageCount for clarity in publications-table-row
KAS-3024/PR pub: v3: reports: columns

Commit:
move and split overdue check up in getter in publication-status + getter in translation/publication-subcase
KAS-3024/PR pub: v3: reports: columns

ghost commented:
- [x] Ik heb de isOverdue-check gedaan op basis van getters in de models. Omdat deze check 2x in de template voorkomt heb ik hem toch in de controller gehouden.
- [x] Statuspills geupdate naar nieuw design.
- [ ] Het herbekijken van het async-getter-pattern is nog niet gebeurd.

Commit:
show publication-status in publications/overview (wating for StatusPill updates)
KAS-3024/PR pub: v3: reports: columns

Commit:
Revert "show publication-status in publications/overview (wating for StatusPill updates)"

This reverts commit 76bd7febc52df3456f83a4909fa5914a48bf6815.

Commit:
show publication-status in publications/overview (wating for StatusPill updates)
KAS-3024/PR pub: v3: reports: columns

Commit:
use StatusPill for publication-status in publications/details (wating for StatusPill updates)
KAS-3024/PR pub: v3: reports: columns

Commit:
restore comment
KAS-3024 pub: v3: reports: columns

erikap committed:
Simplify publication status pill CSS key getter

erikap committed:
Workaround async getter by using 'get' on Proxy object to get attribute

erikap committed:
Simplify get publication date for publication flow helper function

erikap committed:
Add TODO to revisit async getters in the future

Async getters can be done in a better way once the ember-resources
addon can be used

erikap commented:
Ik heb de async getters nog wat herwerkt zodat we geen `await` nodig hebben in de template, maar heb er een TODO bijgezet om te herbijken wanneer we de ember-resources addon kunnen beginnen gebruiken. Maar ik zou de PR daar niet langer op laten blokkeren, want deze houdt ook andere PRs tegen waarin de branch reeds gemerged is (o.a. de tabs story).
Als Jenkins groen is, kan deze gemerged worden.
